AST-based SQLite driver

JanJakes opened the pull request:
Continuing the work from https://github.com/WordPress/sqlite-database-integration/pull/164.

JanJakes committed:
Split, cleanup, and reorganize the SQLite driver prototype

JanJakes committed:
Namespace-prefix the WIP SQLite driver for now to avoid naming conflicts

JanJakes committed:
Copy WP_SQLite_Translator_Tests and run them against the SQLite driver

JanJakes committed:
Copy the core of WP_SQLite_Translator to WP_SQLite_Driver

Additionally, move the current SQLite driver prototype to a temporary class
WP_SQLite_Driver_Prototype to be gradually moved to the new driver.

JanJakes committed:
Add a base generic WP_Parser_Token class, add docs

JanJakes committed:
Complete WP_Parser_Node helper methods, add tests

JanJakes committed:
Add basic support for SELECT statements, add unit tests

JanJakes committed:
Add basic support for INSERT, UPDATE, REPLACE, DELETE

JanJakes committed:
Add basic support for CREATE TABLE, implement data types

JanJakes committed:
Handle system variables

JanJakes committed:
Add support for UPDATE with ORDER BY and LIMIT

JanJakes committed:
Handle specifics of the CREATE TABLE statement

JanJakes committed:
Add basic ALTER TABLE support

JanJakes committed:
Fix MySQL syntax errors in tests

JanJakes committed:
Introduce information schema builder & create information schema tables

JanJakes committed:
Record CREATE TABLE table info in information schema

JanJakes committed:
Record CREATE TABLE column info in information schema

JanJakes committed:
Record CREATE TABLE constraint info in information schema

JanJakes committed:
Record CREATE TABLE inline constraint info in information schema

JanJakes committed:
Sync constraint info to columns table when constraints are modified

JanJakes committed:
Record ALTER TABLE ADD COLUMN(s) in information schema

JanJakes committed:
Record ALTER TABLE ADD CONSTRAINT in information schema

JanJakes committed:
Record ALTER TABLE CHANGE/MODIFY COLUMN in information schema

JanJakes committed:
Record ALTER TABLE DROP COLUMN in information schema

JanJakes committed:
Record ALTER TABLE DROP INDEX in information schema

JanJakes committed:
Execute CREATE TABLE using information schema

JanJakes committed:
Execute ALTER TABLE using information schema

JanJakes committed:
Implement SHOW CREATE TABLE using information schema

JanJakes committed:
Implement SHOW INDEX using information schema

JanJakes committed:
Implement SHOW GRANTS

JanJakes committed:
Implement DESCRIBE using information schema

JanJakes committed:
Ignore index hints

JanJakes committed:
Add support for LIKE BINARY

JanJakes committed:
Add support for REGEXP functions

adamziel committed:
Merge branch 'develop' into ast-sqlite-driver

JanJakes committed:
Add support for LEFT function

JanJakes committed:
Add support for DATE_ADD and DATE_SUB functions

JanJakes committed:
Add support for DATE_FORMAT()

JanJakes committed:
Implement SQL_CALC_FOUND_ROWS and FOUND_ROWS()

JanJakes committed:
Implement FROM DUAL

JanJakes committed:
Implement "CAST(... AS BINARY)"

JanJakes committed:
Add support for CURRENT_TIMESTAMP() and NOW()

JanJakes committed:
Translate datetime literals, fix string literal processing

JanJakes committed:
Add support for HAVING without GROUP BY

JanJakes committed:
Implement DROP for multiple and temporary tables

JanJakes committed:
Implement ON DUPLICATE KEY UPDATE

JanJakes committed:
Implement INSERT IGNORE and UPDATE IGNORE

JanJakes committed:
Implement DROP TABLE in information schema

JanJakes committed:
Implement SHOW TABLE STATUS

JanJakes committed:
Implement SHOW TABLES

JanJakes committed:
Fix saving and displaying index info

JanJakes committed:
Implement ON UPDATE CURRENT_TIMESTAMP

JanJakes committed:
Prefix index names in SQLite to prevent conflicts

JanJakes committed:
Add support for multi-table delete statements

JanJakes committed:
Remove no longer relevant TODOs

JanJakes committed:
Remove no longer used data types cache

JanJakes committed:
Remove destructor

JanJakes committed:
Remove unused variable

JanJakes committed:
Remove unused property

JanJakes committed:
Simplify setting foreign keys pragma

JanJakes committed:
Use AST to detect transactional statements

JanJakes committed:
Convert IFs to switch statement

JanJakes committed:
Remove outdated comment, add TODO

JanJakes committed:
Improve comment

JanJakes committed:
Use get_child_node() calls to make translation unambiguous

JanJakes committed:
Add a TODO comment for AST-querying API

JanJakes committed:
Handle foreign key checks in ALTER TABLE

JanJakes committed:
Improve query method docs and naming

JanJakes committed:
Fix identifier escaping

JanJakes committed:
Throw exceptions for unsupported throw statements

JanJakes committed:
Unify identifier quoting

JanJakes committed:
Improve API of user defined functions

JanJakes committed:
Add a TODO for SQLite client info

JanJakes committed:
Use a unified prefix for internal objects

JanJakes committed:
Fetch column info only for needed columns when adding a constraint

JanJakes committed:
Preserve ROWIDs in ALTER TABLE statements

JanJakes committed:
Fix and improve translation tests

JanJakes committed:
Use correct string literal quotes

JanJakes committed:
Use table schema value in all information schema queries

JanJakes committed:
Remove prototype classes

JanJakes committed:
Remove WIP namespacing (the new classes don't conflict with the existing ones)

JanJakes committed:
Enable usage of PDO in phpcs

JanJakes committed:
Add a feature flag to use the new driver

JanJakes committed:
Use strict SQLite tables to validate data types

See: https://www.sqlite.org/stricttables.html

JanJakes committed:
Store last insert ID

JanJakes committed:
Translate CONCAT function to concat pipes

JanJakes committed:
Correctly translate REPLACE statements

JanJakes committed:
Add basic SHOW VARIABLES support

JanJakes committed:
Add support for IF NOT EXISTS when creating a table

JanJakes committed:
Improve string literal translation

JanJakes committed:
Fix backslash escaping in a test

JanJakes committed:
Fix identifier translation, use backtics due to issues with double quotes

See: https://www.sqlite.org/quirks.html#double_quoted_string_literals_are_accepted

JanJakes committed:
Do not execute statements when there are no IDs to delete

JanJakes committed:
Add basic support for CHAR_LENGTH

JanJakes committed:
Handle null characters in strings

JanJakes committed:
Fix column default value handling

JanJakes committed:
Display info when the AST driver is used

JanJakes committed:
Remove the AST driver files from .gitattributes

JanJakes committed:
Add a test for the reason to use backtick identifier escaping

adamziel commented on admin-page.php:
```diff
@@ -130,7 +130,8 @@ function sqlite_plugin_adminbar_item( $admin_bar ) {
 
 	if ( defined( 'SQLITE_DB_DROPIN_VERSION' ) && defined( 'DB_ENGINE' ) && 'sqlite' === DB_ENGINE ) {
-		$title = '<span style="color:#46B450;">' . __( 'Database: SQLite', 'sqlite-database-integration' ) . '</span>';
+		$suffix = defined( 'WP_SQLITE_AST_DRIVER' ) && WP_SQLITE_AST_DRIVER ? ' (AST)' : '';
```
nice

adamziel commented on grammar-tools/MySQLParser.g4:
```diff
@@ -2991,6 +2991,10 @@ simpleExpr: %simpleExpr_collate (CONCAT_PIPES_SYMBOL %simpleExpr_collate)*;
     | DEFAULT_SYMBOL OPEN_PAR_SYMBOL simpleIdentifier CLOSE_PAR_SYMBOL                                   # simpleExprDefault
     | VALUES_SYMBOL OPEN_PAR_SYMBOL simpleIdentifier CLOSE_PAR_SYMBOL                                    # simpleExprValues
     | INTERVAL_SYMBOL expr interval PLUS_OPERATOR expr                                                   # simpleExprInterval
+    /* @FIX: Move function calls and ref to the end to avoid conflicts with the above expressions. */
```
Does `@FIX` means "to fix"? Or "fixed compared to the original syntax"?

JanJakes replied:
@adamziel This means "fixed in the original Workbench grammar". So far, I've been documenting all changes to the grammar in this way. I don't know if we need to preserve that. There are two main types of changes:

1. Actual fixes — what was wrong or missing in the Workbench grammar. These could be even contributed upstream.
2. Reordering — so that our parser picks the correct path. These are not needed upstream, since ANTLR resolves that when compiling the grammar.

I think maybe for now, we could still preserve it, consider contributing fixes upstream and perhaps also improving our grammar preprocessing so that conflicts are solved automatically. If we do that, these comments will be useful to see where the conflicts appear and to test whether we can fix them more generally.

That said, maybe I should change `FIX` to `FIXED`. What do you think?

adamziel replied:
> That said, maybe I should change FIX to FIXED. What do you think?

Sounds great. Or `CHANGED` or so.

adamziel commented on grammar-tools/MySQLParser.g4:
```diff
@@ -2965,9 +2965,9 @@ simpleExpr: %simpleExpr_collate (CONCAT_PIPES_SYMBOL %simpleExpr_collate)*;
-    | columnRef jsonOperator?                                                                            # simpleExprColumnRef
+    /*| functionCall                                                                                       # simpleExprFunction*/
+    /*| runtimeFunctionCall                                                                                # simpleExprRuntimeFunction*/
+    /*| columnRef jsonOperator?                                                                            # simpleExprColumnRef*/
```
can we remove these commented out lines?

JanJakes replied:
@adamziel These are all related to the `FIX` comments discussed in https://github.com/Automattic/sqlite-database-integration/pull/1#discussion_r1939068451. I think we can remove those when we're ready to remove the fix comments.

adamziel commented on wp-includes/parser/class-wp-parser-node.php:
```diff
@@ -102,83 +102,169 @@ public function merge_fragment( $node ) {
+		return $tokens;
+	}
+
+	public function get_descendants(): array {
```
nit: `get_all_descendant_tokens()` would help differentiate it from `get_descendant_tokens()`

adamziel replied:
actually, do we need a separate method? Maybe `get_descendant_tokens()` already does everything we need?

JanJakes replied:
This actually returns all descendants — both nodes and tokens. It complements `get_children` and it's there to kind of complete the API. The API as a whole also relates to the AST-querying capabilities, as discussed in https://github.com/WordPress/sqlite-database-integration/pull/164#discussion_r1856183561. I think it's still worth exploring in a separate ticket after this initial work is merged.

adamziel replied:
sounds good

JanJakes replied:
I created an issue to explore this further: https://github.com/Automattic/sqlite-database-integration/issues/4

adamziel commented on wp-includes/parser/class-wp-parser.php:
```diff
@@ -52,7 +52,7 @@ private function parse_recursive( $rule_id ) {
 		// possibly match the current token.
 		if ( isset( $this->grammar->lookahead_is_match_possible[ $rule_id ] ) ) {
-			$token_id = $this->tokens[ $this->position ]->type;
+			$token_id = $this->tokens[ $this->position ]->id;
```
I like the change do `token->id`

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+						$err_message = $ex->getMessage();
+					}
+				}
+			} while ( $locked );
```
Do we need a timeout to avoid an infinite loop here? Or, even better, could we trust PDO::ATTR_TIMEOUT and avoid the loop entirely?

JanJakes replied:
@adamziel I'll try to debug this somehow. It would be great to get rid of this loop altogether.

JanJakes replied:
Good news here — the `ATTR_TIMEOUT` seems to be correctly waiting for the lock.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+			}
+		}
+
+		WP_SQLite_PDO_User_Defined_Functions::register_for( $pdo );
```
good naming!

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+
+		WP_SQLite_PDO_User_Defined_Functions::register_for( $pdo );
+
+		// MySQL data comes across stringified by default.
+		$pdo->setAttribute( PDO::ATTR_STRINGIFY_FETCHES, true );
```
what does "stringified" mean? that ints and booleans are strings?

JanJakes replied:
Yes, I'll make it more clear.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+			self::$grammar = new WP_Parser_Grammar( require self::GRAMMAR_PATH );
+		}
+
+		// Fixes a warning in the site-health screen.
```
let's quote the warning in this comment

JanJakes replied:
I documented this in detail in the property definition and added a TODO that it should likely be fixed in WP core.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+		//        consider providing that information lazily via a getter.
+		$this->client_info = SQLite3::version()['versionString'];
+
+		$this->pdo->query( 'PRAGMA foreign_keys = ON' );
```
Is this consistent with how WordPress MySQL tables work by default? I'm thinking about SET modes and MyISAM vs innodb

JanJakes replied:
The default today is InnoDB and in that case, when you add FKs, they just work. In SQLite, however, they need to be turned on. WordPress doesn't use FKs, but some plugins may do so.

adamziel replied:
Let's keep an eye on this – we might need to make it configurable by the caller. WordPress tables may be altered to use MyISAM and plugins may create custom tables with different storage engines. It would be useful to support both strong FK constraints and a relaxed mode.

JanJakes replied:
I added a note about this to a new issue: https://github.com/Automattic/sqlite-database-integration/issues/5

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+			 * Filters queries before they are translated and run.
+			 *
+			 * Return a non-null value to cause query() to return early with that result.
+			 * Use this filter to intercept queries that don't work correctly in SQLite.
```
Why intercept them? Is this for error reporting?

JanJakes replied:
I'm removing these hooks from the new driver for now. We can add them back in the future if needed.

adamziel commented on wp-includes/parser/class-wp-parser-node.php:
```diff
@@ -102,83 +102,169 @@ public function merge_fragment( $node ) {
 
-	public function get_token( $token_id = null ) {
+
+	public function get_child() {
```
Let's call it `get_first_child()` to clarify what happens when there's more than one

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+			}
+
+			if ( $child instanceof WP_Parser_Node && 'simpleStatement' === $child->rule_name ) {
+				$subchild = $child->get_child_node( 'transactionOrLockingStatement' );
```
a few inline comments here and there would improve the readability, but man – the rules names are such a huge help. This reads so much better than the token-based implementation!

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+			return $this->return_value;
+		} catch ( Exception $err ) {
+			// Rollback the nested transaction.
+			$this->rollback();
```
would this throw if an exception is thrown before the first `begin_transaction()`  call?

JanJakes replied:
Good catch, adding a check.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+			$ast    = $parser->parse();
+
+			if ( null === $ast ) {
+				throw new Exception( 'Failed to parse the MySQL query.' );
```
Let's subclass the Exception to avoid mixing up our exceptions with potential exceptions from other plugins (e.g. thrown in hooks and filters)

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+		$ast = $children[0]->get_child_node();
+		switch ( $ast->rule_name ) {
+			case 'selectStatement':
+				$this->query_type = 'SELECT';
```
let's make sure we can handle union selects

JanJakes replied:
I added tests for both UNION and UNION ALL.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+				$this->execute_select_statement( $ast );
+				break;
+			case 'insertStatement':
+				$this->query_type = 'INSERT';
```
Do we need another set of terms? As in, could we use `insertStatement` that we get from the parser instead of introducing `INSERT`?

JanJakes replied:
I removed this property, we're not using it anymore.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+
+		// Handle SQL_CALC_FOUND_ROWS.
+		if ( true === $has_sql_calc_found_rows ) {
+			// Recursively find a query expression with the first LIMIT or SELECT.
```
this was almost impossible to get right in the old driver and now we can just reason about a tree – this is fantastic

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+		 */
+		$where_subquery = null;
+		if ( $has_order || $has_limit ) {
+			$where_subquery = 'SELECT rowid FROM ' . $this->translate_sequence(
```
so good

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+				&& 'updateList' === $child->rule_name
+			) {
+				// We can stop here, as the update statement grammar is:
+				//   ... updateList whereClause? orderClause? simpleLimitClause?
```
thanks for documenting this

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+
+	private function execute_delete_statement( WP_Parser_Node $node ): void {
+		/*
+		 * Multi-table DELETE.
```
so much better than the hardcoded mess from the token-based driver!

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+				return 'AUTOINCREMENT';
+			case WP_MySQL_Lexer::BINARY_SYMBOL:
+				/*
+				 * There is no "BINARY expr" equivalent in SQLite. We can look for
```
Let's make a note that we're not just removing the keyword but we're actually respecting that keyword in the operations where it's used.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+		/*
+		 * 3. Unescape quotes within the string.
+		 */
+		$value = str_replace( $quote . $quote, $quote, $value );
```
Is the tokenizer giving us the literal escaped quotes?

JanJakes replied:
@adamziel It does give the string to us in full, including escape sequences at the moment. That said, it's a very good question! Should it do so? The tokenizer needs to understand escaping anyway (to know when a literal ends), so it could as well process the escape sequences 🤔

adamziel replied:
Sounds like we have good reasons to extract the unescaped data from the tokenizer. Do we also have reasons to extract the escaped data?

Thinking about the future, we could use this parser in WP core to escape prepared statements without round-trips to the MySQL server – in that case, only extracting unescaped content would mean we'd decode and re-encode all the string literals we stumble upon.

Perhaps this behavior could be configurable in the tokenizer, depending on what's more useful for the caller?

JanJakes replied:
> Sounds like we have good reasons to extract the unescaped data from the tokenizer. Do we also have reasons to extract the escaped data?

@adamziel Actually, I'm not sure if we need the unescaped data. What I meant is we probably could move this logic to the tokenizer because the tokenizer needs to look at escaping anyway. I think I will create a separate ticket to verify escaping more in depth and add this idea as well.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+		return implode( $separator, $parts );
+	}
+
+	private function translate_string_literal( WP_Parser_Node $node ): string {
```
cc @dmsnell – there's a lot of escaping happening here, and I know you have deep thoughts and knowledge about MySQL escaping

adamziel replied:
@JanJakes this is the part where things may go very wrong, let's make sure we have a large and comprehensive test suite just for this. Ideally we could just reuse one from some established database driver implementing prepared expressions.

JanJakes replied:
@adamziel As mentioned in https://github.com/Automattic/sqlite-database-integration/pull/1#discussion_r1946737335, I'll create a ticket for this, will add a larger test case, and we can discuss further there.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+			$value = $token->value;
+		}
+
+		return '`' . str_replace( '`', '``', $value ) . '`';
```
Is doubling the backticks the only rule we have to follow? Is there a way to derail this? What would happen if there's a backslash backtick combination? Or something like a LTR UTF-8 character inside the string?

JanJakes replied:
I believe so, yes. In SQLite literals, there are no Unicode escape sequences or anything like that. You can only escape the quote, and only by doubling it. A few quick tests:

```sql
SELECT 'a\b'; -- a\b
SELECT 'a'b'; -- syntax error
SELECT 'a\'b'; -- syntax error
SELECT 'a\''b'; -- a\'b
```

And here is a quick test with LTR:

```sql
SELECT 'a‎b' == 'ab'; -- 0 (because the first literal has LTR between 'a' and 'b')
```

adamziel replied:
That sounds promising, thanks! Just to be extra diligent here – is there anything in the SQLite test suite related to escaping that catches your attention?

JanJakes replied:
@adamziel As mentioned in https://github.com/Automattic/sqlite-database-integration/pull/1#discussion_r1946737335, I'll create a ticket for this, will add a larger test case, and we can discuss further there.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2592 @@
+	 * It is executed only once when the installation begins.
+	 */
+	private function prepare_directory() {
+		$u = umask( 0000 );
```
let's document inline what is the consequence of calling umask().

adamziel commented:
Monumental work here @JanJakes 🎉  You're amazing. This is it – the solid, methodically crafted database driver that enables WordPress to use other databases. It's a profound moment for WordPress, thank you so much for all your hard work here ❤️ 

I've left some inline notes. Let's give escaping the special attention it needs. Also, I left a bunch of other minor comments. That being said, I am **very** happy with this PR. Thank you thank you thank you!

JanJakes committed:
Use PDO to load SQLite version, improve docs

JanJakes committed:
Remove all remaining usages of SQLite 3

JanJakes committed:
Declare PDO and PDO SQLite requirements in composer.json

JanJakes committed:
Use getters instead of public properties

JanJakes committed:
Fix query dumping script

JanJakes committed:
Simplify SQLite lock wait logic

JanJakes committed:
Remove legacy hooks

JanJakes committed:
Closely mirror wpdb::query() and wpdb::_do_query()

JanJakes committed:
Simplify driver state

JanJakes committed:
Remove uselsess return values

JanJakes committed:
Change FIX comments to CHANGED

JanJakes committed:
Improve parser node method naming

JanJakes committed:
Fix FROM/IN handling for SHOW statements

JanJakes committed:
Add tests for UNION and UNION ALL

JanJakes committed:
Configure the SQLite driver using an $options array

JanJakes committed:
Use custom exception class for driver errors

JanJakes committed:
Ignore transaction rollback errors when an exception occurs

JanJakes committed:
Return true for transactional commands

JanJakes committed:
Move database directory creation outside of the driver, revamp it

JanJakes committed:
Extract debug mode setting to an option

JanJakes committed:
Improve driver docs and property naming

JanJakes committed:
Revamp error handling, extract error notice creation out from the driver

JanJakes committed:
Improve docs and naming

JanJakes committed:
Check for mimimum SQLite version

JanJakes commented:
@adamziel Thanks for the review! I think I addressed most of the issues and [created tickets](https://github.com/Automattic/sqlite-database-integration/issues) for the remaining ones. Please, let me know if I missed anything.

adamziel commented on composer.json:
```diff
@@ -8,7 +8,9 @@
 	},
 	"require": {
-		"php": ">=7.0"
+		"php": ">=7.0",
```
7.2+, so same as WordPress trunk, is also fine. No strong preferences, just saying

JanJakes replied:
I wasn't sure to which version to bump to, so I created a ticket (and we discussed it in Slack): https://github.com/Automattic/sqlite-database-integration/issues/12

So I think we can bump it as part of that ticket...

adamziel commented on grammar-tools/MySQLParser.g4:
```diff
@@ -153,12 +153,12 @@ alterStatement:
         | alterServer
         // ALTER USER is part of the user management rule.
-        | alterInstance /* @FIX: Add support for "ALTER INSTANCE ..." statement. */
+        | alterInstance /* @CHANGED: Add support for "ALTER INSTANCE ..." statement. */
```
Perfect, this `@CHANGED` annotation is super clear

adamziel replied:
Maybe past tense for the comment? So "Added support"?

JanJakes replied:
Good idea, fixed.

adamziel commented on health-check.php:
```diff
@@ -31,7 +31,7 @@ function sqlite_plugin_filter_debug_data( $info ) {
 		$info['wp-database']['fields']['database_version'] = array(
 			'label' => __( 'SQLite version', 'sqlite-database-integration' ),
-			'value' => class_exists( 'SQLite3' ) ? SQLite3::version()['versionString'] : null,
+			'value' => $info['wp-database']['fields']['server_version'] ?? null,
```
nice

adamziel commented on wp-includes/parser/class-wp-parser-node.php:
```diff
@@ -102,83 +102,169 @@ public function merge_fragment( $node ) {
+
+	/*
+	 * @TODO: Let's implement a more powerful AST-querying API.
+	 *        See: https://github.com/WordPress/sqlite-database-integration/pull/164#discussion_r1855230501
```
If we're refactoring the API, let's annotate each method we expect to change as `@private` or `@internal` and add a comment for the API consumers explaining the reasons not to use them yet. Otherwise, we'll introduce breaking changes without a prior notice. Maybe we could `@deprecate` them right away – that way the IDEs will provide visual hints.

JanJakes replied:
@adamziel Maybe, while we're behind the feature flag and have a bunch of `TODO`s, we could still consider all the new APIs to be experimental and subject to change? Given we're under the Automattic org repo, maybe that would be enough? Or do you think it's safer to annotate all public methods for now?

adamziel replied:
It could be a separate ticket, we's just have to resolve it before handing the api over for consumption by another team

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2601 @@
+	 *   https://www.sqlite.org/lang_datefunc.html
+	 *   https://strftime.org/
+	 */
+	const DATE_FORMAT_TO_STRFTIME_MAP = array(
```
Nitpick: I noticed I needed to read the comment to understand the mapping, we could encode more information in the constant name:

```suggestion
		const MYSQL_DATE_FORMAT_TO_SQLITE_STRFTIME_MAP = array(
```

JanJakes replied:
Good point! Fixed.

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2601 @@
+					}
+
+					if (
+						WP_MySQL_Lexer::ROLLBACK_SYMBOL === $token1->id
```
Should we bale out if we see `ROLLBACK TO SAVEPOINT` until we can support that? Same for creating savepoints.

adamziel replied:
Actually we do support them internally so maybe baling out would be as easy as preserving the savepoint semantics

JanJakes replied:
Ticket: https://github.com/Automattic/sqlite-database-integration/issues/14

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2601 @@
+				$alias_map[ $alias ] = $ref;
+			}
+
+			// 3. Compose the SELECT query to fetch ROWIDs to delete.
```
Any chance we'll also see HAVING, GROUP BY, LEFT JOIN etc.? It's fine if we do, I'd just bale out if we see any unsupported criterion.

More broadly, it would be fantastic to have a catalog of things we do support (a whitelist/allowlist) and compare the incoming query against it rather than assume we know all the relevant parts of the incoming query and we can always cherry-pick the what's relevant. It could be as simple as a foreach loop that must make a bale/ignore/preserve/transform decision about every single node it sees. Thinking out loud, it could have a `switch` statement with a catch-all `default` case that bales out on anything we don't expect to see.

The mental model would be a version or "innocent until proven guilty" – "all queries error out unless they're explicitly supported".

JanJakes replied:
> Any chance we'll also see HAVING, GROUP BY, LEFT JOIN etc.? It's fine if we do, I'd just bale out if we see any unsupported criterion.

Luckily, none of these is possible in a MySQL DELETE statement. That said, we don't have any specific handling of `DELETE` with `LIMIT` and SQLite [seems to support it only with a specific compile flag](https://www.sqlite.org/lang_delete.html#optional_limit_and_order_by_clauses).

> More broadly, it would be fantastic to have a catalog of things we do support (a whitelist/allowlist)...

Do you mean the DELETE statement in particular, or more like a general rule? I'd like to be more restrictive in general and steer the driver more in the allowlist direction overall, but it can get pretty complicated — a foreach loop may not work very well, since nodes can be deeply nested, have different meanings in different context, order may change the meaning, etc.

That said, I do think we could steer it a bit more towards the "allowlist" approach overall, but it may be a different approach on a case-by-case basis, and maybe it's not absolutely necessary for all cases (e.g., SELECT and other read statements). Most importantly, all DDL statements should be allowlisted, so that we know what we support on the information schema side and so that the information schema data cannot become incorrect.

Considering how large this PR is and that at this stage it's still behind a feature flag, what do you think if I create tickets for these points and describe all of this? I'd like to get this initial PR done so that we can then work in smaller, more focused chunks.

adamziel replied:
Sure, separate issues seem fine for this. And sure, if a strict allowlist isn't an option then let's examine realistic options that move us in the right direction. Thank you for all the great work here!

JanJakes replied:
I created a ticket: https://github.com/Automattic/sqlite-database-integration/issues/15

adamziel commented on wp-includes/sqlite-ast/class-wp-sqlite-driver.php:
```diff
@@ -0,0 +1,2601 @@
+			return;
+		}
+
+		// @TODO: Translate DELETE with JOIN to use a subquery.
```
Making an explicit decision about every AST node would a) make it apparent to the code reader b) prevent running a DELETE query with a condition different from what the developer supplied (and potentially leading to data loss)

JanJakes replied:
I guess this is discussed in [the comment above](https://github.com/Automattic/sqlite-database-integration/pull/1#discussion_r1952132444) — I agree that we should go more in the "allowlist" direction, at least for DDL and write statements, but the particular implementation may differ on a case-by-case basis, and I'd create a separate ticket for that, if that makes sense to you.

JanJakes committed:
Improve constant name

JanJakes committed:
Use past tense in @CHANGED comments

adamziel commented:
Great work @janjakes! There are specific directions to go from here, but regardless - this is such a huge improvement ❤️
